yabt/buildcontext.py

```diff
@@ -181,6 +183,93 @@ def get_target_extraction_context(self, build_file_path: str) -> dict:
     #     """Register a named BuildEnv Docker image in this build context."""
     #     self.buildenv_images[name] = docker_image
 
+    def get_buildenv_graph(self):
+        """Return a graph induced by buildenv nodes"""
+        buildenvs = set(target.buildenv for target in self.targets.values()
+                        if target.buildenv)
+        return nx.DiGraph(self.target_graph.subgraph(reduce(
+            lambda x, y: x | set(y),
+            (get_descendants(self.target_graph, buildenv)
+             for buildenv in buildenvs), buildenvs)))
+
+    def ready_nodes_iter(self, graph_copy):
+        """Generate ready targets from the graph `graph_copy`.
+
+        The input graph is mutated by this method, so it has to be a mutable
+        copy of the graph (e.g. not original copy, or read-only view).
+        """
+
+        def is_ready(target_name):
```

Comment:
this means that you are checking if target name is ready. As Avi explained to me last week, this approach is wrong (I've made the same mistake). When a node is done, it clears a counter on all it's descendants. At which point you check what this counter is. At which point you know if to put it in the ready queue. This "bottom up" approach is much faster than the top down approach.